File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
def solution(places):
    answer = []
    
    for place in places:
        place = [list(i) for i in place]
        check = True
        
        for y in range(5):
            if not check:
                break
            for x in range(5): # 우하단만 체크
                if place[y][x] == 'P':
                    if x < 4:
                        if place[y][x+1] == 'P':
                            check = False
                        elif place[y][x+1] == 'O' and (x < 3 and place[y][x+2] == 'P'):
                            check = False
                    if y < 4:
                        if place[y+1][x] == 'P':
                            check = False
                        elif place[y+1][x] == 'O' and (y < 3 and place[y+2][x] == 'P'):
                            check = False
                    if x < 4 and y > 0:
                        if place[y-1][x+1] == 'P' and (place[y][x+1] == 'O' or place[y-1][x] == 'O'):
                            check = False
                    if x < 4 and y < 4:
                        if place[y+1][x+1] == 'P' and (place[y][x+1] == 'O' or place[y+1][x] == 'O'):
                            check = False
        if check:
            answer.append(1)
        else:
            answer.append(0)
    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;
class Solution {
    int[][] dir;
    public int[] solution(String[][] places) {
        int[] answer = new int[] {1,1,1,1,1};
        dir = new int[][]{{1,0},{-1,0},{0,1}};
        for(int i=0; i<5; i++){
            loop:for(int j=0; j<5; j++){
                for(int k=0; k<5; k++){
                    if(places[i][j].charAt(k) == 'P'){
                        if(bfs(places[i], j, k)){
                            answer[i] = 0;
                            break loop;
                        }
                    }
                }
            }
        }
        return answer;
    }
    public boolean bfs(String[] place, int y, int x){
        boolean[][] visit = new boolean[5][5];
        Queue<int[]> q = new ArrayDeque<>();
        visit[y][x] = true;
        q.add(new int[] {y,x,0});
        while(!q.isEmpty()){
            int[] out = q.remove();
            int cy = out[0], cx = out[1], dep = out[2];
            if(dep > 0 && place[cy].charAt(cx) == 'P'){
                return true;
            }
            if(dep < 2){
                for(int[] d : dir){
                    int ny = cy + d[0], nx = cx + d[1];
                    if(0<=ny && ny<5 && 0<=nx && nx<5 && visit[ny][nx] == false && place[ny].charAt(nx) != 'X'){
                        visit[ny][nx] = true;
                        q.add(new int[] {ny,nx,dep+1});
                    }
                }
            }
        }
        return false;
    }
}